MTG data structure

imane-aanna opened the pull request.

imane-aanna committed:
Fixes openalea-incubator/csMTG#43

- Defines the new data structure.
- Tests the constructor.

imane-aanna committed:
Fixes openalea-incubator/csMTG#45

- Added the 4 functions that are related to the attribute scale.
- Added tests to verify the results.

imane-aanna committed:
Fixes openalea-incubator/csMTG#46

Added functions to calculate the number of vertices per scale and list
them.
Added the corresponding tests.

imane-aanna committed:
Fixes openealea-incubator/csMTG#43

Added functions which iterate on edges at a given scale.
Added the corresponding tests.

imane-aanna committed:
Fixes openalea-incubator/csMTG#48

Added functions that enable to list the siblings of a vertex and count
them.
Added the corresponding tests.

imane-aanna committed:
Fixes openalea-incubator/csMTG#49

Added functions that enable to list all roots of the MTG depending on
the scale.

imane-aanna committed:
Fixes openalea-incubator/csMTG#50

Added a function that enables to return the complex a vertex.
Added a test.

imane-aanna commented on code/mtg.cs:
```diff
@@ -237,6 +241,48 @@ public List<int> Roots(int scale = 0)
+        /// <param name="vertexId"> Vertex identifier. </param>
+        /// <returns> The complex of the vertex or null. </returns>
+        public int? Complex(int vertexId)
+        {
```
@pradal : In case the vertex doesn't have a complex, should it return null or a default value ?

pradal commented:
@imane-aanna : Return the same thing that you return if a vertex do not have a parent.

imane-aanna committed:
Fixes openalea-incubator/csMTG#51

Added accessors to the attribute components, as well as a function which
iterates the roots of the graphs composing the vertex, as well as a
counter of the number of components.
Still need to add the tests, but can't do so unless the traversal
functions are updated.

imane-aanna committed:
Partially fixes openalea-incubator/csMTG#53

Added the function ComplexAtScale.

imane-aanna committed:
Fixes openalea-incubator/csMTG#52

Updates the traversal functions so as to take into the parameters an mtg
instead of a PropertyTree, and also add a new optional parameter:
"complex".

imane-aanna committed:
Closes openalea-incubator/csMTG#51

Added the missing public function : ComponentRoots.

imane-aanna committed:
Reorganize the tree file

Put regions in the tree file so as to gather functions which operate on
the same attribute or have the same purpose.

imane-aanna committed:
Fixes openalea-incubator/csMTG#55

- Added function InsertParent for Tree, PropertyTree, Mtg.
- Added a test for InsertParent on Tree.
- Fixes problems with the functions: ReplaceParent, AddChild, NewChild,
RemoveVertex.
- Modified NbChildren, Children.
- Changed the privacy of the function NewId().

imane-aanna committed:
Fixes openalea-incubator/csMTG#56

- Added the function InsertSibling for the 3 data structures.
- Tested it for the Tree, and the PropertyTree.

imane-aanna committed:
Fixes openalea-incubator/csMTG#57

- Added function AddComponent that enables to add a component to a
complex.
- Added the corresponding tests.

imane-aanna committed:
Tests openalea-incubator/csMTG#51

- Tests functions related to the attribute: "components".
- Fixed a small issue with the traversal. (Not all nodes have an
edge_type when created).

imane-aanna committed:
Closes openalea-incubator/csMTG#53

Added the missing function: ComponentsAtScale.

imane-aanna committed:
Fixes openalea-incubator/csMTG#58

- Added the function that enables to a child and a complex.
- Added the corresponding tests.

imane-aanna committed:
Fixes openalea-incubator/csMTG#59

Added the function that enables to add a tree to a parent in another
tree.

imane-aanna committed:
Fixes openalea-incubator/csMTG#60

- Added the function Clear() for all classes (Tree, PropertyTree, MTG).
- Added the corresponding tests.

imane-aanna committed:
Fixes openalea-incubator/csMTG#61

Added a function SimpleTree which can generate and add a regular tree to
an existing one at a given vertex.
Added the corresponding tests.

imane-aanna committed:
Fixes openalea-incubator/csMTG#62 and openalea-incubator/csMTG#63

- Added the function RemoveVertex for the MTG.
- Added the corresponding tests.
- Update the RandomTree generator so that it adds a random tree to a
vertex.
- RandomTree now returns the identifier of the last vertex added.

imane-aanna committed:
Fixes openalea-incubator/csMTG#64

Added a function that lists the vertices composing the tree.

imane-aanna committed:
Fixes openalea-incubator/csMTG#65

Added the function IsLeaf

imane-aanna committed:
Inverse the parameters of ReplaceParent

imane-aanna committed:
Fixes openalea-incubator/csMTG#66

Added the function RandomMtg and all the other functions it calls, for
instance:
- ComputeMissingEdges
- LookForCommonAncestors
- FullAncestors
- FatMtg
- ColoredTree

imane-aanna committed:
Modifications on the RandomMtg

imane-aanna committed:
Fixes openalea-incubator/csMTG#67

Added a function that displays a tree.
It still needs one line to be completed (I don't know what the tab[:-1]
means).

imane-aanna committed:
Fixes openalea-incubator/csMTG#69

Added a function that enables to display the roots of the graph that
composes a vertex at a specific scale. (Added the iterator as well).

imane-aanna committed:
Fixes openalea-incubator/csMTG#68

Added a function that enables to display an MTG.
Added functions that enable to iterate the vertices of an mtg and its
scales (MtgIterator and ScaleIterator).

imane-aanna committed:
Closes openalea-incubator/csMTG#67

Got the signification of tab=tab[:-1] => Equivalent of pop (All items
are kept except the last one).

imane-aanna committed:
Fixes openalea-incubator/csMTG#70

- Added the function subtree.
- Added a default parameter of root in the tree's constructor.

imane-aanna committed:
Fixes openalea-incubator/csMTG#70

Added the function InsertSiblingTree

imane-aanna committed:
Fixes openalea-incubator/csMTG#72

Added the function: AddChildTree for the class Tree.

imane-aanna committed:
Fixes openalea-incubator/csMTG#73

Added a function that removes all vertices starting from a vertex in the
parameters.

imane-aanna committed:
Invert the order of the parameters for InsertSibling

imane-aanna committed:
Changed the order of the parameters in InsertParent

imane-aanna committed:
Fixes openalea-incubator/csMTG#74

- Added the function SubTree.
- Added the root accessor.

imane-aanna committed:
Fixes openalea-incubator/csMTG#75

- Added the function InsertSiblingTree for class PropertyTree.

imane-aanna committed:
Fixes openalea-incubator/csMTG#76

- Added the function AddChildTree for the class PropertyTree.
- Fixed a small issue in InsertSiblingTree. (Confused two parameters).

imane-aanna committed:
Fixes openalea-incubator/csMTG#77

- Added the function RemoveVertex for PropertyTree.
- Changed the return value of RemoveVertex in the class Tree

imane-aanna committed:
Add comments

Added comments for the function AddComponent
Added the cases were the scale wouldn't already exist
